Objects/frameobject.c

```diff
@@ -150,109 +156,144 @@ markblocks(PyCodeObject *code_obj, int len)
                 case POP_JUMP_IF_FALSE:
                 case POP_JUMP_IF_TRUE:
                 case JUMP_IF_NOT_EXC_MATCH:
-                    j = get_arg(code, i);
+                {
+                    int j = get_arg(code, i);
                     assert(j < len);
-                    if (blocks[j] == -1 && j < i) {
+                    if (blocks[j] == UNINITIALIZED && j < i) {
                         todo = 1;
                     }
-                    assert(blocks[j] == -1 || blocks[j] == block_stack);
-                    blocks[j] = block_stack;
+                    if (opcode == JUMP_IF_NOT_EXC_MATCH) {
+                        block_stack = pop_block(block_stack);
+                    }
+                    int64_t target_stack = block_stack;
+                    if (opcode != JUMP_IF_FALSE_OR_POP && opcode != JUMP_IF_TRUE_OR_POP) {
+                        target_stack = pop_block(target_stack);
+                    }
+                    block_stack = pop_block(block_stack);
+                    assert(blocks[j] == UNINITIALIZED || blocks[j] == target_stack);
+                    blocks[j] = target_stack;
```

Comment:
FWIW, the split here with `target_stack` is a little hard to follow at first.  A slight change would help with that:

```suggestion
                    int64_t target_stack = block_stack;
                    if (opcode == JUMP_IF_NOT_EXC_MATCH) {
                        block_stack = pop_block(block_stack);
                        target_stack = block_stack;
                    }
                    if (opcode != JUMP_IF_FALSE_OR_POP && opcode != JUMP_IF_TRUE_OR_POP) {
                        target_stack = pop_block(target_stack);
                    }
                    block_stack = pop_block(block_stack);
                    assert(blocks[j] == UNINITIALIZED || blocks[j] == target_stack);
                    blocks[j] = target_stack;
```

...though it still feels off somehow.  Perhaps a comment would be worth adding here to make sure it is clear what is happening?

Reply:
Your updates make this section perfectly clear now.  Thanks!